zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanStore.java

```diff
@@ -468,21 +456,101 @@
               .setUUID("end_ts", UUIDs.endOf(endTsMillis))
               .setInt("limit_", limit);
 
-      return transform(session.executeAsync(bound),
-          new Function<ResultSet, Map<TraceIdUDT, Long>>() {
-            @Override public Map<TraceIdUDT, Long> apply(ResultSet input) {
-              Map<TraceIdUDT, Long> traceIdsToTimestamps = new LinkedHashMap<>();
-              for (Row row : input) {
-                traceIdsToTimestamps.put(
-                        row.get("trace_id", TraceIdUDT.class),
-                        row.getLong("ts"));
-              }
-              return traceIdsToTimestamps;
-            }
-          }
-      );
+      return transform(session.executeAsync(bound), readResultsAsMap(Collections.emptyMap(), traceIdToLong));
     } catch (RuntimeException ex) {
       return immediateFailedFuture(ex);
     }
   }
+
+  private static <K, T> AsyncFunction<ResultSet, Map<K, T>> readResultsAsMap(final Map<K, T> results,
+      Function<Row, Map.Entry<K, T>> rowMapper) {
+    return new AsyncFunction<ResultSet, Map<K, T>>() {
+      @Override public ListenableFuture<Map<K, T>> apply(ResultSet rs) throws Exception {
+        // How far we can go without triggering the blocking fetch:
+        int remainingInPage = rs.getAvailableWithoutFetching();
```

Comment:
@michaelsembwever thanks for taking a look at this.  I'm not sure I completely follow your suggestion though.  Is there a reference example that you can point me at?

The approach I've taken here was borrowed from Datastax documentation on async paging https://docs.datastax.com/en/developer/java-driver/3.1/manual/async/#async-paging